lib/source-destination/balena-s3-source.ts

```diff
@@ -27,42 +28,111 @@ import { ZipSource } from './zip';
 
 type Name = 'balena' | 'resin';
 
-export class BalenaS3Source extends SourceDestination {
-	private rawSource: Http;
-	private zipSource: ZipSource;
-	private ready: Promise<void>;
-	private names: Name[] = ['balena', 'resin'];
-	public name: Name; // images can be named balena.img or resin.img
+export interface AwsCredentials {
+	accessKeyId: string;
+	secretAccessKey: string;
+}
 
-	constructor(
-		readonly bucket: string,
-		readonly deviceType: string,
-		readonly version: string,
-		readonly host = 's3.amazonaws.com',
-	) {
-		// example:
-		// bucket: resin-staging-img or resin-production-img-cloudformation
-		// deviceType: raspberry-pi
-		// version: 2.9.6+rev1.prod
+export interface BalenaS3SourceOptions {
+	host?: string; // https://s3.amazonaws.com
+	bucket?: string; // resin-staging-img or resin-production-img-cloudformation
+	prefix?: string; // images or preloaded-images or esr-images
+	deviceType: string; // raspberry-pi
+	buildId: string; // 2.9.6+rev1.prod
+	release?: string; // 1344795
+	awsCredentials?: AwsCredentials;
+}
+
+export abstract class BalenaS3SourceBase extends SourceDestination {
+	public readonly host: string;
+	public readonly bucket: string;
+	public readonly prefix: string;
+	public readonly deviceType: string;
+	public readonly buildId: string;
+	public readonly release?: string; // Only used for preloaded images
+	protected axiosInstance: AxiosInstance;
+	private static filesMissingFromPreloadedImages = [
+		'VERSION',
+		'VERSION_HOSTOS',
+		'device-type.json',
+	];
+
+	constructor({
+		host = 'https://s3.amazonaws.com',
+		bucket = 'resin-production-img-cloudformation',
+		prefix = 'images',
```

Comment:
Can you please remove these defaults? These options are already defaulted at an earlier point and I'd rather we ensured these values were explicitly pushed down here.

Reply:
Removed